core/src/ids.rs

```diff
@@ -204,10 +205,10 @@ impl ProgramId {
     pub const SYSTEM: Self = Self(*b"geargeargeargeargeargeargeargear");
 
     /// Generate ProgramId from given CodeId and salt
-    pub fn generate(code_id: CodeId, salt: &[u8]) -> Self {
+    pub fn generate<T: ProgramNonce>(nonce: T, code_id: CodeId, salt: &[u8]) -> Self {
         const SALT: &[u8] = b"program";
 
-        let argument = [SALT, code_id.as_ref(), salt].concat();
+        let argument = [nonce.as_nonce().as_ref(), SALT, code_id.as_ref(), salt].concat();
```

Comment:
overhead out of nowhere, because your as_nonce allocates vec just to use it as slice :/

Reply:
The only way I found of doing this was by using 

```unsafe { slice::from_raw_parts(self as *const u32 as *const u8, 4) }```

That allows me to return a &[u8]. This is still safe since we know that the u32 should still live after the bytes are returned thus it will be a valid pointer for when it is added to the Hash. Please confirm that this is okay, or some other alternative solution to removing the Vec.

Reply:
That's mostly about my misunderstanding why do we need trait here at all